posix: Update to new kernel timeout API

pfalcon opened the pull request:
Based on https://github.com/zephyrproject-rtos/zephyr/issues/24926.

pfalcon committed:
lib: posix: Update to new timeout API

Mostly trivial search-and-replace, except for pthread_rwlock.c, where
we need spread timeout over 2 semaphore operations.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
tests: posix: common: Switch k_sleep() -> k_msleep().

As a step in migration to new timeout API.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
tests: posix: eventfd: Don't mix POSIX and Zephyr kernel APIs

K_MSEC() shouldn't be used with poll(), as they come from 2 different
API domains.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

carlescufi commented on lib/posix/pthread_rwlock.c:
```diff
@@ -71,7 +71,7 @@ int pthread_rwlock_rdlock(pthread_rwlock_t *rwlock)
 	}
 
-	return read_lock_acquire(rwlock, K_FOREVER);
+	return read_lock_acquire(rwlock, -1);
```
Use `SYS_FOREVER_MS`

carlescufi commented on lib/posix/pthread_mutex.c:
```diff
@@ -78,7 +78,7 @@ int pthread_mutex_timedlock(pthread_mutex_t *m,
 {
 	s32_t timeout = (s32_t)timespec_to_timeoutms(abstime);
-	return acquire_mutex(m, timeout);
+	return acquire_mutex(m, K_MSEC(timeout));
```
Use `SYS_TIMEOUT_MS()` instead which will convert `SYS_FOREVER_MS` to `K_FOREVER` automatically.

pfalcon replied:
Specifically here there's no either SYS_FOREVER_MS or K_FOREVER.

carlescufi replied:
I see, thanks

carlescufi commented on lib/posix/pthread_rwlock.c:
```diff
@@ -215,8 +215,19 @@ int pthread_rwlock_unlock(pthread_rwlock_t *rwlock)
+	 * that a smart compiler may optimize it out (for a common definition
+	 * of K_FOREVER.
+	 */
+	if (timeout == -1) {
```
use `SYS_TIMEOUT_MS` instead of rewriting the logic

pfalcon replied:
> use SYS_TIMEOUT_MS instead of rewriting the logic

Ack, for this file it indeed makes sense.

carlescufi commented on lib/posix/pthread_rwlock.c:
```diff
@@ -231,17 +242,32 @@ static u32_t write_lock_acquire(pthread_rwlock_t *rwlock, s32_t timeout)
+	 * that a smart compiler may optimize it out (for a common definition
+	 * of K_FOREVER.
+	 */
+	if (timeout == -1) {
```
Use `SYS_TIMEOUT_MS`

carlescufi commented on lib/posix/pthread_rwlock.c:
```diff
@@ -231,17 +242,32 @@ static u32_t write_lock_acquire(pthread_rwlock_t *rwlock, s32_t timeout)
-		if (timeout > K_NO_WAIT) {
+	if (k_sem_take(&rwlock->wr_sem, k_timeout) == 0) {
+		/* update remaining timeout time for 2nd sem */
+		if (timeout == -1) {
```
Consider using `SYS_TIMEOUT_MS` and in any case use `SYS_FOREVER_MS` to compare against

carlescufi commented on tests/posix/common/src/mutex.c:
```diff
@@ -28,7 +28,7 @@ void *normal_mutex_entry(void *p1)
 			break;
 		}
-		k_sleep(SLEEP_MS);
+		k_sleep(K_MSEC(SLEEP_MS));
```
You can use `k_msleep()` instead

pfalcon replied:
Done

pfalcon replied:
>> You can use k_msleep() instead

> Done

No, wait, k_sleep has prototype k_sleep(k_timeout_t timeout), but k_msleep - k_msleep(s32_t ms)? But that makes no sense!

andyross replied:
k_msleep() and k_usleep() are simple wrappers that take integers in specific  units; they work like APIs in other systems.  There was desire last summer that k_sleep() take a k_timeout_t to allow it to use multiprecision and absolute arguments (so people who wanted that wouldn't have to use a k_timer, basically).

The suggestion was to change that code to `k_msleep(SLEEP_MS);`

carlescufi commented on tests/posix/common/src/mutex.c:
```diff
@@ -102,7 +102,7 @@ void test_posix_normal_mutex(void)
 		TC_PRINT("Thread1 creation failed %d", ret);
 	}
-	k_sleep(SLEEP_MS);
+	k_sleep(K_MSEC(SLEEP_MS));
```
You can use `k_msleep()` instead

carlescufi commented on tests/posix/eventfd/src/main.c:
```diff
@@ -152,7 +152,7 @@ static void test_eventfd_poll_event(void)
 	pfd.events = POLLIN;
 
-	ret = poll(&pfd, 1, K_SECONDS(3));
+	ret = poll(&pfd, 1, 3000);
```
`3 * MSEC_PER_SEC` ?

pfalcon replied:
Nope, just time in milliseconds, as accepted by POSIX poll().

carlescufi reviewed (changes requested):
Thank you @pfalcon! A few comments but this looks almost ready

pfalcon committed:
tests: posix: clock: Workaround to align clocks before querying

Before calling clock_gettime() 2 times in row, issue k_usleep(1)
to align code execution to timer interrupt to prevent (well,
minimize) possibility of getting different ticks values.

Suggested by @andyross.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Updated to use SYS_TIMEOUT_MS() where there was clear duplication of what it does.

pfalcon commented on lib/posix/pthread_mutex.c:
```diff
@@ -20,7 +20,7 @@ static const pthread_mutexattr_t def_attr = {
 };
 
-static int acquire_mutex(pthread_mutex_t *m, int timeout)
+static int acquire_mutex(pthread_mutex_t *m, k_timeout_t timeout)
```
@carlescufi: So, the only way I can productively interpret comment "Use SYS_TIMEOUT_MS() instead which will convert SYS_FOREVER_MS to K_FOREVER automatically." below (outside pthread_rwlock.c, where it clearly makes sense) is as follows: "You can skip patching all the aux functions to accept k_timeout_t, instead can keep that at int/s32_t, and just use SYS_TIMEOUT_MS() inside such aux funcs to convert to k_timeout_t at the last mile."

If that's what you suggest, then:
1. How would I know that?
2. I specifically don't know that, as the whole idea seemed to be to use k_timeout_t as the primary timeout type. And I performed this refactor with that in mind. Then what you suggest is effectively along the lines of "you can ignore and workaround that k_timeout_t nuisance for as long as possible", so I'm not sure how good is it.

carlescufi replied:
No, I was only suggesting it in the places that needed a conversion from `s32_t` to `k_timeout_t` and where `-1` is a parameter, I wasn't suggesting that you stay with `s32_t`. Maybe I didn't express myself clearly enough, sorry about that.

pfalcon replied:
> No, I was only suggesting it in the places that needed a conversion from s32_t to k_timeout_t and where -1 is a parameter, I wasn't suggesting that you stay with s32_t. 

Good, then the current state of this PR should be exactly like that.

pfalcon commented:
This is now includes a workaround proposed in https://github.com/zephyrproject-rtos/zephyr/pull/25000#issuecomment-624776808 and expected to pass CI.

@carlescufi, please re-review.

pfalcon commented:
@carlescufi : Now passes CI.

carlescufi reviewed (approved):
This looks good to me now, thanks

carlescufi commented:
@andyross could we get an additional approval from you?

andyross reviewed (approved):
This all looks good.

And yeah: forgotten tick alignment has been a pervasive antipattern through basically all our test code (much of it mine, too).  Timing is hard.  RTOS developers tend to assume their code is in a critical section and that steps have been taken to lock it against external change.  But these APIs don't work like that, because you can't lock the passage of time.
